HW2/P5/physics.pyx

```diff
@@ -55,55 +76,97 @@ cdef void sub_update(FLOAT[:, ::1] XY,
                      float R,
                      int i, int count,
                      UINT[:, ::1] Grid,
-                     float grid_spacing) nogil:
+                     float grid_spacing,
+                     int grid_size,
+                     omp_lock_t *locks) nogil:
     cdef:
         FLOAT *XY1, *XY2, *V1, *V2
         int j, dim
+        int XYG[2]
         float eps = 1e-5
+        int gsx, gsy
+        int gs = 3 # the size of the circle we are going to search
 
     # SUBPROBLEM 4: Add locking
+    acquire(&locks[i])
```

Comment:
Locks thread i on the outside, and j on the inside for loop. Locks are increasing order so no deadlocking occurs.